File: src/main/java/com/leon/bugreport/API/ErrorClass.java
package com.leon.bugreport.API;

import com.leon.bugreport.logging.ErrorMessages;

import java.io.File;
import java.io.FileWriter;

import static com.leon.bugreport.BugReportManager.plugin;

public class ErrorClass {
	private static final File LOG_DIR = new File("plugins/BugReport/logs");
	private static final File ERROR_FILE = new File(LOG_DIR, "error.log");

	public static void logErrorMessage(String message) {
		int errorTries = 0;
		while (errorTries < 3) {
			try {
				logErrorMessageInternal(message);
				return;
			} catch (Exception e) {
				errorTries++;
			}
		}
	}

	private static void logErrorMessageInternal(String message) {
		cacheDirectoryExists();
		createErrorLog();

		try {
			appendToFile(message);
		} catch (Exception e) {
			String errorMessage = ErrorMessages.getErrorMessage(11);
			plugin.getLogger().warning(errorMessage);
			logErrorMessage(errorMessage);
		}
	}

	private static void cacheDirectoryExists() {
		if (!LOG_DIR.exists() && !LOG_DIR.mkdirs()) {
			String errorMessage = ErrorMessages.getErrorMessage(12);

			plugin.getLogger().warning(errorMessage);
			logErrorMessage(errorMessage);
		}
	}

	public static void createErrorLog() {
		try {
			if (!ERROR_FILE.exists() && !ERROR_FILE.createNewFile()) {
				String errorMessage = ErrorMessages.getErrorMessage(13);

				plugin.getLogger().warning(errorMessage);
				logErrorMessage(errorMessage);
			}
		} catch (Exception ignored) {
		}
	}

	private static void appendToFile(String error) {
		try {
			String currentTime = java.time.LocalDateTime.now().toString();
			String formattedTime = currentTime.substring(0, currentTime.indexOf('T')) + ' ' + currentTime.substring(currentTime.indexOf('T') + 1, currentTime.indexOf('.'));

			FileWriter fileWriter = new FileWriter(ErrorClass.ERROR_FILE, true);
			fileWriter.write('[' + formattedTime + "] " + error + '\n');
			fileWriter.close();
		} catch (Exception e) {
			String errorMessage = ErrorMessages.getErrorMessage(14);

			plugin.getLogger().warning(errorMessage);
			logErrorMessage(errorMessage);
		}
	}
}

File: src/main/java/com/bugreportmc/bugreport/api/ErrorClass.kt
package com.bugreportmc.bugreport.api

import com.bugreportmc.bugreport.BugReportManager
import com.bugreportmc.bugreport.BugReportPlugin.Companion.plugin
import java.io.File
import java.io.FileWriter
import java.time.LocalDateTime

object ErrorClass {
	private val LOG_DIR = File("plugins/BugReport/logs")
	private val ERROR_FILE = File(LOG_DIR, "error.log")

	@JvmStatic
	fun logErrorMessage(errorMessage: String) {
		cacheDirectoryExists()
		createErrorLog()

		try {
			appendToFile(errorMessage)
		} catch (e: Exception) {
			plugin.logger.warning("Failed to write to error log file")
			logErrorMessage("Failed to write to error log file")
		}
	}

	private fun cacheDirectoryExists() {
		if (!LOG_DIR.exists() && !LOG_DIR.mkdirs()) {
			plugin.logger.warning("Failed to create cache directory")
			logErrorMessage("Failed to create cache directory")
		}
	}

	private fun createErrorLog() {
		try {
			if (!ERROR_FILE.exists() && !ERROR_FILE.createNewFile()) {
				plugin.logger.warning("Failed to create error log file")
				logErrorMessage("Failed to create error log file")
			}
		} catch (ignored: Exception) {
		}
	}

	private fun appendToFile(error: String) {
		try {
			val currentTime = LocalDateTime.now().toString()
			val formattedTime = currentTime.substring(
				0, currentTime.indexOf('T')
			) + ' ' + currentTime.substring(currentTime.indexOf('T') + 1, currentTime.indexOf('.'))

			val fileWriter = FileWriter(ERROR_FILE, true)
			fileWriter.write("[$formattedTime] $error\n")
			fileWriter.close()
		} catch (e: Exception) {
			plugin.logger.warning("Failed to write to error log file")
			logErrorMessage("Failed to write to error log file")
		}
	}
}
